apps/zk-explorer/utils/index.ts

```diff
@@ -8,35 +13,181 @@ export function createProjectDetailPath(
 
   if (!trimmedUrl.startsWith(GITHUB_URL_PREFIX)) {
-    return false;
+    return null;
   }
 
-  const significantPathSegment = trimmedUrl.slice(GITHUB_URL_PREFIX.length);
-  const parts = significantPathSegment.split('/');
+  const slugPathSegment = trimmedUrl.slice(GITHUB_URL_PREFIX.length);
+  const parts = slugPathSegment.split('/');
 
-  // The significant path segment should be in the form of
+  // The slug path segment should be in the form of
   // <username>/<repository-name>, so there should be exactly
   // two parts.
   if (parts.length !== 2) {
-    return false;
+    return null;
   }
 
   const segmentRegex = /^([a-zA-Z0-9_-]+)$/i;
   const owner = parts[0];
   const repo = parts[1];
 
-  return segmentRegex.test(owner) && segmentRegex.test(repo);
+  if (!segmentRegex.test(owner) || !segmentRegex.test(repo)) {
+    return null;
+  }
+
+  return [owner, repo];
 }
 
-export function tryOrElse<T>(fn: () => T, fallback: () => T): T {
+export function validateGithubUrl(url: string): boolean {
+  return parseGithubUrl(url) !== null;
+}
+
+// TODO: This is temporary, until the backend is implemented.
+export function getMockProjects(): ProjectSearchResponseData {
+  const mockProjects = Array<ProjectItem>(ITEMS_PER_PAGE).fill({
+    ownerAvatarUrl:
+      'https://avatars.githubusercontent.com/u/76852793?s=200&v=4',
+    repositoryOwner: 'webb',
+    repositoryName: 'masp',
+    stargazerCount: 123,
+    circuitCount: 24,
+    description:
+      'Short blurb about what the purpose of this circuit. This is a longer line to test multiline.',
+    contributorAvatarUrls: Array(15).fill(
+      'https://avatars.githubusercontent.com/u/76852793?s=200&v=4'
+    ),
+  });
+
+  return {
+    projects: mockProjects,
+    resultCount: mockProjects.length,
+  };
+}
+
+// TODO: This is temporary, until the backend is implemented.
+export function getMockCircuits(): CircuitSearchResponseData {
+  const mockCircuits = Array<CircuitItem>(ITEMS_PER_PAGE).fill({
+    ownerAvatarUrl:
+      'https://avatars.githubusercontent.com/u/76852793?s=200&v=4',
+    filename: 'circuit.circom',
+    description:
+      'Short blurb about what the purpose of this circuit. This is a longer line to test multiline.',
+    stargazerCount: 123,
+    constraintCount: 456,
+  });
+
+  return {
+    circuits: mockCircuits,
+    resultCount: mockCircuits.length,
+  };
+}
+
+export function setSearchParam(key: SearchParamKey, value: string | null) {
+  const updatedSearchParams = new URLSearchParams(window.location.search);
+
+  if (value === null || value === '') {
+    updatedSearchParams.delete(key);
+  } else {
+    updatedSearchParams.set(key, value);
+  }
+
+  const searchParamsString =
+    updatedSearchParams.size === 0 ? '' : `?${updatedSearchParams}`;
+
+  const newUrl = window.location.pathname + searchParamsString;
+
+  window.history.pushState({}, '', newUrl);
+}
```

Comment:
I suggest setting the search parameters using Next.js functions instead of directly modifying the native browser URL. This approach is preferable as it can help avoid potential bugs related to UI rendering when there is a connection between search parameters and the UI. You can find more information about this in the following discussion: https://github.com/vercel/next.js/discussions/47583.

Reply:
I've added `nuqs` to the dependencies, which allows me to accomplish just that via a `useQueryState` hook.